### PR TITLE
Adding new global agent's configurations 

### PR DESCRIPTION
**Pull request by pereyra-m**
|Related issue|
|---|
|#6310|

## Description

In this PR, two new global configurations were added:
- _agents_disconnection_alert_time_: it is used by **monitord** to create an agent disconnection alert
- _agents_disconnection_time_: it is the internal time to consider an agent disconnected, and it is used by **remoted** to avoid the communication with agents that are no longer available.

Also, these global configurations can be requested to **monitord** and **remoted** sending a "getconfig global" message to the corresponding queue.
## Configuration options
- _agents_disconnection_alert_time_: default value 20 seconds (s,m,d,h can be used)
- _agents_disconnection_time_: default value 120 seconds (s,m,d,h can be used)
## Tests

The manual tests performed are detailed in the issue.

- Compilation without warnings in every supported platform
  - [x] Linux
- [x] Source installation
- [ ] Review logs syntax and correct language

- Memory tests for Linux
  - [x] Scan-build report
  - [x] Valgrind (memcheck and descriptor leaks check)

### CODE REVIEW

File: src/monitord/main.c [hunk 1]
```diff
@@ -162,6 +162,13 @@ int main(int argc, char **argv)
         merror_exit(CONFIG_ERROR, cfg);
     }
 
+    /* Setting default values and reading global configuration */
+    Config.agents_disconnection_time = 20;
+    Config.alert_agent_disconnection_time = 2;
```
Comment (DProvinciani) [inline src/monitord/main.c:167]: This should be saved in seconds

File: src/config/global-config.h [hunk 2]
```diff
@@ -111,6 +111,9 @@ typedef struct __Config {
     int min_rotate_interval;
     ssize_t max_output_size;
     long queue_size;
+
+    int agents_disconnection_time;
+    int alert_agent_disconnection_time;
```
Comment (palaciosjeremias) [inline src/config/global-config.h:116]: Ditto

File: src/monitord/monitor_agents.c [hunk 3]
```diff
@@ -20,7 +20,7 @@ void monitor_agents()
     char **cr_agents;
     char **av_agents;
 
-    av_agents = get_agents_by_last_keepalive(GA_ACTIVE, mond.delete_old_agents > 0 ? mond.delete_old_agents * 60 : DISCON_TIME);
+    av_agents = get_agents_by_last_keepalive(GA_ACTIVE, mond.delete_old_agents > 0 ? mond.delete_old_agents * 60 : Config.alert_agent_disconnection_time * 60);
```
Comment (DProvinciani) [inline src/monitord/monitor_agents.c:23]: This computation of `*60` should be removed after changing the readding of settings.

File: src/monitord/monitord.c [hunk 4]
```diff
@@ -157,6 +156,18 @@ cJSON *getMonitorInternalOptions(void) {
     return root;
 }
 
+cJSON *getMonitorGlobalOptions(void) {
+
+    cJSON *root = cJSON_CreateObject();
+    cJSON *monconf = cJSON_CreateObject();
+
+    cJSON_AddNumberToObject(monconf,"agents_disconnection_time",Config.agents_disconnection_time);
+    cJSON_AddNumberToObject(monconf,"alert_agent_disconnection_time",Config.alert_agent_disconnection_time);
```
Comment (DProvinciani) [inline src/monitord/monitord.c:165]: In this part of the code, considering that the setting will be saved in memory as seconds, we should add to the JSON `Config.alert_agent_disconnection_time/60` (using the name of the setting suggested in the first comment)

File: src/remoted/main.c [hunk 5]
```diff
@@ -127,13 +125,20 @@ int main(int argc, char **argv)
         merror_exit(CONFIG_ERROR, cfg);
     }
 
+    /* Setting default values and reading global configuration */
+    Config.agents_disconnection_time = 20;
+    Config.alert_agent_disconnection_time = 2;
```
Comment (DProvinciani) [inline src/remoted/main.c:130]: Use seconds.

File: src/monitord/monitor_agents.c (hunk 3 above)
Comment (DProvinciani) [reply src/monitord/monitor_agents.c:23]: Also, remember that the `alert_agent_disconnection_time` is a time that should pass after `agents_disconnection_time` to generate an alert. So, the right computation to generate an alert here should be:

```
agents_disconnection_time + alert_agent_disconnection_time
```

**Review by DProvinciani — CHANGES_REQUESTED** [review]
Please, apply changes according to the comments.

File: src/monitord/main.c [hunk 6]
```diff
@@ -162,6 +162,13 @@ int main(int argc, char **argv)
...
+    if (ReadConfig(CGLOBAL, cfg, &Config, NULL) < 0) {
```
Comment (palaciosjeremias) [inline src/monitord/main.c:168]: Cant't we use an unique config struct with multiple masks like analysisd does?
c |= CREPORTS
c |= CGLOBAL

File: src/remoted/main.c [hunk 7]
```diff
@@ -127,13 +125,20 @@ int main(int argc, char **argv)
...
+    if (ReadConfig(CGLOBAL, cfg, &Config, NULL) < 0) {
```
Comment (palaciosjeremias) [inline src/remoted/main.c:131]: This should be part of RemotedConfig() method, including default values.

File: src/monitord/main.c [hunk 8]
```diff
@@ -162,6 +162,13 @@ int main(int argc, char **argv)
         merror_exit(CONFIG_ERROR, cfg);
     }
 
+    /* Setting default values and reading global configuration */
+    Config.agents_disconnection_time = 20;
```
Comment (palaciosjeremias) [inline src/monitord/main.c:166]: We should create a method to set default values and read the configuration, as in Remoted.

**Review by palaciosjeremias — CHANGES_REQUESTED** [review]

File: src/config/global-config.c [hunk 9]
```diff
@@ -150,6 +150,9 @@ int Read_Global(XML_NODE node, void *configp, void *mailp)
     const char *xml_custom_alert_output = "custom_alert_output";
     const char *xml_rotate_interval = "rotate_interval";
     const char *xml_max_output_size = "max_output_size";
+    const char *xml_agents_disconnection_time = "agents_disconnection_time";
+    const char *xml_alert_agent_disconnection_time = "alert_agent_disconnection_time";
```
[the remark this answers is not in the document]
Comment (pereyra-m) [reply src/config/global-config.c:154]: The name has been updated!

File: src/config/global-config.c [hunk 10]
```diff
@@ -653,6 +656,34 @@ int Read_Global(XML_NODE node, void *configp, void *mailp)
...
+                }
+
+            }
+        } else if (strcmp(node[i]->element, xml_alert_agent_disconnection_time) == 0) {
+            if (Config) {
+                if (!OS_StrIsNum(node[i]->content)) {
+                    merror(XML_VALUEERR, node[i]->element, node[i]->content);
+                    return (OS_INVALID);
+                }
+                Config->alert_agent_disconnection_time = atoi(node[i]->content);
```
[the remark this answers is not in the document]
Comment (pereyra-m) [reply src/config/global-config.c]: Now both options are save in seconds.

File: src/monitord/main.c (hunk 1 above)
Comment (pereyra-m) [reply src/monitord/main.c]: Now both options are saved in seconds.

File: src/monitord/monitor_agents.c (hunk 3 above)
Comment (pereyra-m) [reply src/monitord/monitor_agents.c]: Thank you!
This has been fixed.

File: src/monitord/monitord.c (hunk 4 above)
Comment (pereyra-m) [reply src/monitord/monitord.c]: Now the user can use a char (_s,m,h,d_) to indicate the unit, but it is safer to return the time in seconds.

File: src/remoted/main.c (hunk 5 above)
Comment (pereyra-m) [reply src/remoted/main.c]: Done!

File: src/config/global-config.c [hunk 11]
```diff
@@ -150,6 +150,9 @@ int Read_Global(XML_NODE node, void *configp, void *mailp)
     const char *xml_custom_alert_output = "custom_alert_output";
     const char *xml_rotate_interval = "rotate_interval";
     const char *xml_max_output_size = "max_output_size";
+    const char *xml_agents_disconnection_time = "agents_disconnection_time";
```
[the remark this answers is not in the document]
Comment (pereyra-m) [reply src/config/global-config.c:153]: Now the parameters are called: _agents_disconnection_alert_time_ and _agents_disconnection_time_

File: src/config/global-config.c [hunk 12]
```diff
@@ -653,6 +656,34 @@ int Read_Global(XML_NODE node, void *configp, void *mailp)
...
+            }
+        } else if (strcmp(node[i]->element, xml_alert_agent_disconnection_time) == 0) {
+            if (Config) {
+                if (!OS_StrIsNum(node[i]->content)) {
+                    merror(XML_VALUEERR, node[i]->element, node[i]->content);
+                    return (OS_INVALID);
+                }
+                Config->alert_agent_disconnection_time = atoi(node[i]->content);
+
+                if (Config->alert_agent_disconnection_time < 2) {
```
[the remark this answers is not in the document]
Comment (pereyra-m) [reply src/config/global-config.c]: Done!

File: src/config/global-config.h (hunk 2 above)
Comment (pereyra-m) [reply src/config/global-config.h]: Fixed!

File: src/monitord/main.c (hunk 6 above)
Comment (pereyra-m) [reply src/monitord/main.c]: Now that **mond** has a __Config global_ struct field, a second call to **ReadConfig()** is necessary, just like analysisd does

File: src/remoted/main.c (hunk 7 above)
Comment (pereyra-m) [reply src/remoted/main.c]: Done!

File: src/monitord/main.c (hunk 8 above)
Comment (pereyra-m) [reply src/monitord/main.c]: Now, the _Config global_ field is read at the same time that **mond**, but a new method was not created.

File: src/config/global-config.c [hunk 13]
```diff
@@ -654,6 +657,80 @@ int Read_Global(XML_NODE node, void *configp, void *mailp)
                 }
 
             }
+        }
+        /*Agent's disconnection time parameters */
```
Comment (palaciosjeremias) [inline src/config/global-config.c:661]: I think this comment should be replicated at least on the other new config. But it can be deleted.

Comment (palaciosjeremias) [inline src/config/global-config.c:661]: Code style: Use white space after /*

File: src/monitord/main.c [hunk 14]
```diff
@@ -156,9 +156,15 @@ int main(int argc, char **argv)
     mond.emailfrom = NULL;
     mond.emailidsname = NULL;
 
+    /* Setting default agent's global configuration */
+    mond.global.agents_disconnection_time = 20;
```
Comment (palaciosjeremias) [inline src/monitord/main.c:160]: We can move this initialization, together with the previous ones, and ReadConfig to a monitord config method, as in other modules. Like RemotedConfig()

File: src/config/global-config.c [hunk 15]
```diff
@@ -654,6 +657,80 @@ int Read_Global(XML_NODE node, void *configp, void *mailp)
...
+        else if (strcmp(node[i]->element, xml_agents_disconnection_time) == 0) {
+            if (Config) {
+                char c;
+
+                switch (sscanf(node[i]->content, "%d%c", &Config->agents_disconnection_time, &c)) {
```
Comment (DProvinciani) [inline src/config/global-config.c:666]: Shouldn't we implement this time conversion logic in a separate method? We are repeating the same code within the same method.
Ideally, this should be a logic available in a helper method for all the Wazuh daemons.

File: src/config/global-config.c (hunk 13 above)
Comment (DProvinciani) [reply src/config/global-config.c:661]: +1

File: src/monitord/main.c (hunk 14 above)
Comment (DProvinciani) [reply src/monitord/main.c:160]: +1

**Review by DProvinciani — CHANGES_REQUESTED** [review]
Please, apply changes according to the comments.

File: src/config/global-config.c (hunk 13 above)
Comment (pereyra-m) [reply src/config/global-config.c:661]: It was replicated.
They are just to improve clarity.

Comment (pereyra-m) [reply src/config/global-config.c:661]: Fixed!

File: src/monitord/main.c (hunk 14 above)
Comment (pereyra-m) [reply src/monitord/main.c]: Now there is a **MonitordConfig()** that reads the configuration

File: src/config/global-config.c (hunk 15 above)
Comment (pereyra-m) [reply src/config/global-config.c]: That method already existed!
Now both settings use **w_parse_time()**

**Review by DProvinciani — APPROVED** [review]
Looks good!

**Review by palaciosjeremias — APPROVED** [review]
LGTM

**Pull request merged**